src/steamship/agents/functions_based.py

```diff
@@ -0,0 +1,127 @@
+        return self.PROMPT
+
+    def next_action(self, context: AgentContext) -> Action:
+        # Build the Chat History that we'll provide as input to the action
+        messages = build_chat_history(self.default_system_message(), self.message_selector, context)
+        # get working history (completed actions)
+        messages.extend(self._function_calls_since_last_user_message(context))
+
+        # Run the default LLM on those messages
+        output_blocks = self.llm.generate(messages=messages, capabilities=self.capabilities)
+
+        for block in output_blocks:
+            if block.mime_type == MimeTypes.STEAMSHIP_PLUGIN_FUNCTION_CALL_INVOCATION:
+                invocation = FunctionCallingSupport.FunctionCallInvocation.from_block(block)
+                tool = self.tools_map.get(invocation.tool_name)
+                if tool is None:
+                    raise SteamshipError(
+                        f"LLM attempted to invoke tool {invocation.tool_name}, but {self.__class__.__name__} does not have a tool with that name."
+                    )
+                # TODO (PR): Could overload Action for this case to piggy back Invocations on it, but ehhhhhhhhh...
+                # TODO Block parse for input.  text is the default argument that we currently pass in for Tools.  As
+                #  part of a refactor to allow for other parameters, this would need to change.
+                future_action = Action(
+                    tool=tool.name, input=[Block(text=invocation.args.get("text"))], output=None
+                )
+                break
+        else:
+            future_action = FinishAction()
+            invocation = None
+        if not isinstance(future_action, FinishAction):
+            # record the LLM's function response in history
+            assert invocation
+            self._record_function_invocation(invocation, context)
+        return future_action
+
+    def _function_calls_since_last_user_message(self, context: AgentContext) -> Iterable[Block]:
+        function_calls = []
+        for block in context.chat_history.messages[::-1]:  # is this too inefficient at scale?
+            if block.chat_role == RoleTag.USER:
+                return reversed(function_calls)
+            if get_tag(block.tags, kind=TagKind.ROLE, name=RoleTag.FUNCTION):
+                function_calls.append(block)
+            elif get_tag(block.tags, kind=TagKind.FUNCTION_SELECTION):
+                function_calls.append(block)
+        return reversed(function_calls)
+
+    def _record_function_invocation(
+        self, invocation: FunctionCallingSupport.FunctionCallInvocation, context: AgentContext
+    ):
+        tags = [
+            Tag(
+                kind=TagKind.CHAT,
+                name=ChatTag.ROLE,
+                value={TagValueKey.STRING_VALUE: RoleTag.ASSISTANT},
+            ),
+            Tag(kind=TagKind.FUNCTION_SELECTION, name=invocation.tool_name),
+            Tag(
```

Comment:
we can drop this tag now, as `request-id` tagging should be automatic in next release of Engine.